cfgrammar/src/lib/yacc/ast.rs

```diff
@@ -202,8 +209,138 @@ impl GrammarAST {
                 spans: vec![Span::new(0, 0)],
             });
         }
+
+        for sym in &self.expect_unused {
+            match sym {
+                crate::yacc::ast::Symbol::Rule(sym_name, sym_span) => {
+                    if self.get_rule(sym_name).is_none() {
+                        return Err(YaccGrammarError {
+                            kind: YaccGrammarErrorKind::UnknownRuleRef(sym_name.clone()),
+                            spans: vec![*sym_span],
+                        });
+                    }
+                }
+                crate::yacc::ast::Symbol::Token(sym_name, sym_span) => {
+                    if !self.has_token(sym_name) {
+                        return Err(YaccGrammarError {
+                            kind: YaccGrammarErrorKind::UnknownToken(sym_name.clone()),
+                            spans: vec![*sym_span],
+                        });
+                    }
+                }
+            }
+        }
         Ok(())
     }
+
+    pub(crate) fn unused_symbol_warnings(&self) -> impl Iterator<Item = YaccGrammarWarning> {
+        #[derive(Hash, PartialEq, Eq, Debug, Copy, Clone)]
+        enum SymbolKind {
+            Rule,
+            Token,
+        }
+        let start_rule_name = self.start.as_ref().map(|(name, _)| name.clone());
+        let start_rule = self
+            .rules
+            .iter()
+            .find(|(rule_name, _)| start_rule_name.as_ref() == Some(rule_name));
+        let mut used_symbols = HashSet::new();
+        let mut expect_unused = self
+            .expect_unused
+            .iter()
+            .filter_map(|sym| match &sym {
+                Symbol::Rule(sym_name, _) => {
+                    if self.rules.contains_key(sym_name) {
+                        Some((sym_name, SymbolKind::Rule))
+                    } else {
+                        None
+                    }
+                }
+                Symbol::Token(sym_name, _) => {
+                    if self.tokens.contains(sym_name) {
+                        Some((sym_name, SymbolKind::Token))
+                    } else {
+                        None
+                    }
+                }
+            })
+            .collect::<Vec<(&String, SymbolKind)>>();
+        // If a rule is specified in `%expect-unused`, also add the tokens associated with it.
```

Comment:
Hmm, I think there are 2 kinds of changes that can cause unused token errors to appear here 

1. When you remove a rule `%expect-unused` while it is still unreachable, you now would get an errors for the rule and the tokens.
2. When you remove a token from a `%expect-unused` unreachable rule, while it is still used by another unreachable rule not present in `%expect-unused`.  You'd get 1 new error.

I'm pretty certain at least that actually fixing an error here can only cause (zero or more) errors to be fixed, but never introduced, since making the rule reachable makes the tokens in it reachable.

Are these cases similar enough to the situations you described?

Reply:
I went ahead and removed this in [edf85b9](https://github.com/softdevteam/grmtools/pull/345/commits/edf85b973db8fac2b7d23fc50da5a7ea59a8bceb), I'm definitely in the for doing something to reduce the number of warnings here, as it seems like perhaps token rules caused by unused rules *seem* superfluous and noisy in many cases.

But either way, it is better to start off simple, and we can always reduce the noise later after we've gained some experience with using it.